Test gen strings update regression

allredj opened the pull request:
Update tests and add smoke tests to the regression test suite.
Depends on https://github.com/diffblue/cbmc/pull/906.

thk123 commented on regression/strings/smoke_tests/performance.py:
```diff
@@ -0,0 +1,27 @@
+
+file_name='performance_'+commit_id+'.out'
+print "writing to file", file_name
+f=open(file_name, 'w')
```
Prefer:
```python
with open(file_name, 'w') as f:
    for x in process.split('\n'):
       f.write(x+'\n')
```

allredj replied:
Done.

smowton commented on regression/strings/smoke_tests/performance.py:
```diff
@@ -0,0 +1,27 @@
+
+commit_id=time.strftime("%Y_%m_%d__%H_%M_%S")+"__"+commit[7:]
+
+process = check_output(["grep", "^Runtime\ decision\ procedure", "-R"])
```
Might be wrong but I don't think those escaped spaces are doing anything (as in, Python will already submit "Runtime decision procedure" as a single argument, and grep will regard "\ " and " " as the same AFAIK)

allredj replied:
Escapes removed.

smowton commented on regression/strings/smoke_tests/performance.py:
```diff
@@ -0,0 +1,27 @@
+print "writing to file", file_name
+f=open(file_name, 'w')
+
+for x in process.split('\n'):
```
Does this differ from just piping grep to that file? Try using
```
with open("/tmp/crap", "w") as f:
  subprocess.check_call(..., stdout=f)
```

Note use of `check_call` not `check_output`, as the latter would try to capture `stdout` to a string.

allredj replied:
I'm not sure what you mean here. I can't grep on the folder where `f` is currently being written, I need to buffer somehow. But I removed the `split` which is unnecessary.

smowton commented on regression/strings/smoke_tests/performance.py:
```diff
@@ -0,0 +1,27 @@
+
+gnuplot_output = check_output(["gnuplot", "-e", "file='"+file_name+"'", "-e", "outputfile='"+file_name+".png'", "performance_draw.gp"])
+
+print gnuplot_output
```
In that case, use `check_call` instead of `check_output`, which just permits `gnuplot` to write to stdout as usual.

allredj replied:
Done.

allredj commented:
I corrected a problem in the `strings-smoke-tests/Makefile` that made the tests fail with Appveyor. I also rebased, which allowed to enable the test for `String.contains()` (which has just been merged: https://github.com/diffblue/cbmc/pull/828). The corrections on the Python script are upcoming.

allredj commented on regression/strings/smoke_tests/performance.py:
```diff
@@ -0,0 +1,27 @@
+import time
+from subprocess import check_output
+
+git_output=check_output(["git", "show", "HEAD"])
```
Done.

allredj commented on regression/strings/smoke_tests/performance.py:
```diff
@@ -0,0 +1,27 @@
+process = check_output(["grep", "^Runtime\ decision\ procedure", "-R"])
+
+file_name='performance_'+commit_id+'.out'
+print "writing to file", file_name
```
Corrected.

allredj commented on regression/strings/smoke_tests/performance.py:
```diff
@@ -0,0 +1,27 @@
+
+print "drawing to file", file_name+".png"
+
+gnuplot_output = check_output(["gnuplot", "-e", "file='"+file_name+"'", "-e", "outputfile='"+file_name+".png'", "performance_draw.gp"])
```
Done.

allredj commented on regression/strings/smoke_tests/performance.py:
```diff
@@ -0,0 +1,27 @@
+gnuplot_output = check_output(["gnuplot", "-e", "file='"+file_name+"'", "-e", "outputfile='"+file_name+".png'", "performance_draw.gp"])
+
+print gnuplot_output
+
```
Removed.

allredj commented on regression/strings-smoke-tests/java_append_int/test.desc:
```diff
@@ -1,4 +1,4 @@
-CORE
+FUTURE
```
These tests have never actually been enabled, because the test suite was not called from the `regression/Makefile`. For this one in particular, the behaviour might be affected when we add models for strings. It is just easier to go through all these tests once everything is merged, then enable those that work, and create issues for those that don't. There is no particular issue to enable the tests, but it is an ongoing process.

allredj commented on regression/strings-smoke-tests/java_append_int/test.desc:
```diff
@@ -1,4 +1,4 @@
-CORE
+FUTURE
 test_append_int.class
 --refine-strings
```
I understand that `--string-max-length something` will become the default on test-gen. However, on cbmc itself, the option should not be necessary (because we do not need to create real counterexamples). If anything, the option should be removed in the test descriptions where it is present, but that is not a priority I think.

allredj commented:
The python script has been beautified following the reviewer's request. It has also been PEP8'ed.

smowton commented on regression/strings-smoke-tests/performance.py:
```diff
@@ -0,0 +1,34 @@
+commit = git_output.split('\n')[0]
+commit_id = time.strftime('%Y_%m_%d__%H_%M_%S')+'__'+commit[7:]
+
+process = check_output(['grep', '^Runtime decision procedure', '-R'])
```
Change this to something like

```
file_name = 'performance_'+commit_id+'.out'
print 'writing to file', file_name
with open(file_name, 'w') as f:
  check_call(['grep', '^Runtime decision procedure', '-R'], stdout=f)
```

Otherwise lgtm

allredj replied:
@smowton As suggested above, this doesn't work because we cannot write the output of a grep in the directory we are currently grepping. I get the following error:
```
writing to file performance_2017_06_05__14_53_24__56800d52eca33b06978f8863f333ef43fbabc6c3.out
grep: input file 'performance_2017_06_05__14_53_24__56800d52eca33b06978f8863f333ef43fbabc6c3.out’ is also the output
Traceback (most recent call last):
  File "performance.py", line 29, in <module>
    check_call(['grep', '^Runtime decision procedure', '-R'], stdout=f)
  File "/usr/lib/python2.7/subprocess.py", line 541, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['grep', '^Runtime decision procedure', '-R']' returned non-zero exit status 2
```

smowton replied:
Ah, makes sense

thk123 commented on regression/strings-smoke-tests/performance.py:
```diff
@@ -0,0 +1,34 @@
+process = check_output(['grep', '^Runtime decision procedure', '-R'])
+file_name = 'performance_'+commit_id+'.out'
+
+print 'writing to file', file_name
```
If you're fixing this, doesn't this also need the comment here: https://github.com/diffblue/cbmc/pull/976#discussion_r120071368

(i.e. `print('writing to file ' + file_name)`

thk123 commented on regression/strings-smoke-tests/performance.py:
```diff
@@ -0,0 +1,34 @@
+with open(file_name, 'w') as f:
+    f.write(process)
+
+print('drawing to file', file_name+'.png')
```
This still using the comma/python2 thing so if you agree I think this should become:

```python
print('drawing to file '+file_name+'.png')
```
